### PR TITLE
Add Task.WaitAll overload for `IEnumerable<Task>`

### LINKED ISSUE #34106: Task.WaitAll missing IEnumerable<Task> overload
In general, `Task.WhenAll` is used to wait for a collection for tasks to complete execution. However, `Task.WaitAll` still has valid niche use-cases. For example, in a synchronous method that blocks shutdown until it completes - `appLifetime.ApplicationStopping.Register(() => _shutdownTokenSource.Cancel())`.

`Task.WaitAll` seems to be missing an overload that takes an `IEnumerable<Task>`. This overload exists for `Task.WhenAll`. Can we add this?

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/runtime/issues/34106

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs [hunk 1]
```diff
@@ -4831,18 +4845,48 @@ public static void WaitAll(Task[] tasks, CancellationToken cancellationToken)
...
-            if (tasks == null)
+            if (tasks is null)
             {
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.tasks);
             }
+
+            ReadOnlySpan<Task> span =
+                tasks is List<Task> list ? CollectionsMarshal.AsSpan(list) :
+                tasks is Task[] array ? array :
+                CollectionsMarshal.AsSpan(new List<Task>(tasks));
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:4880]: subjective nit: `switch` looks better for such constructs

```suggestion
            ReadOnlySpan<Task> span = tasks switch
            {
                List<Task> list => CollectionsMarshal.AsSpan(list),
                Task[] array => array,
                _ => CollectionsMarshal.AsSpan(new List<Task>(tasks))
            };
```

File: src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs [hunk 2]
```diff
@@ -4831,18 +4845,48 @@ public static void WaitAll(Task[] tasks, CancellationToken cancellationToken)
...
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> to observe while waiting for the tasks to complete.</param>
+        /// <exception cref="ArgumentNullException">The <paramref name="tasks"/> argument is null.</exception>
+        /// <exception cref="ArgumentException">The <paramref name="tasks"/> argument contains a null element.</exception>
+        /// <exception cref="ObjectDisposedException">One or more of the <see cref="Task"/> objects in tasks has been disposed.</exception>
+        /// <exception cref="OperationCanceledException">The <paramref name="cancellationToken"/> was canceled.</exception>
+        /// <exception cref="AggregateException">
+        /// At least one of the <see cref="Task"/> instances was canceled. If a task was canceled, the <see cref="AggregateException"/>
+        /// contains an <see cref="OperationCanceledException"/> in its <see cref="AggregateException.InnerExceptions"/> collection.
+        /// </exception>
         [UnsupportedOSPlatform("browser")]
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:4869]: the other public overloads contain following annotation:

```suggestion
        [UnsupportedOSPlatform("browser")]
        [MethodImpl(MethodImplOptions.NoOptimization)] // this is needed for the parallel debugger
```

why this one does not?

**Review by adamsitnik — APPROVED** [review]
LGTM, I am looking forward to the removal of `.ToArray` in all the cases where users were constructing a `List<Task>` before calling the pre-existing overloads! (BTW should we create an analyzer for that?)

Before I hit the merge button please let me know whether lack of `[MethodImpl(MethodImplOptions.NoOptimization)]` was intentional or not.

Thanks!

**Comment by adamsitnik** [comment]
I am going to squash it now so it gets included in P4. In case there is a need for the attribute, we can always add it later (but I expect there is no such need as this method is more complex compared to other overloads that most likely got inlined in the past)

File: src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:4869]: @gregg-miskelly, are these NoOptimizations in Task code still needed by the debugger?

**Pull request merged**